e2e_tests/tests/all_providers/cross.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use e2e_tests::TestClient;
+use parsec_client::core::interface::requests::{ProviderID, Result};
+
+const HASH: [u8; 32] = [
+    0x69, 0x3E, 0xDB, 0x1B, 0x22, 0x79, 0x03, 0xF4, 0xC0, 0xBF, 0xD6, 0x91, 0x76, 0x37, 0x84, 0xA2,
+    0x94, 0x8E, 0x92, 0x50, 0x35, 0xC2, 0x8C, 0x5C, 0x3C, 0xCA, 0xFE, 0x18, 0xE8, 0x81, 0x37, 0x78,
+];
+
+const PLAINTEXT_MESSAGE: [u8; 32] = [
+    0x69, 0x3E, 0xDB, 0x1B, 0x22, 0x79, 0x03, 0xF4, 0xC0, 0xBF, 0xD6, 0x91, 0x76, 0x37, 0x84, 0xA2,
+    0x94, 0x8E, 0x92, 0x50, 0x35, 0xC2, 0x8C, 0x5C, 0x3C, 0xCA, 0xFE, 0x18, 0xE8, 0x81, 0x37, 0x78,
+];
+
+fn setup_sign(provider: ProviderID, key_name: String) -> Result<(TestClient, Vec<u8>, Vec<u8>)> {
+    let mut client = TestClient::new();
+    client.set_provider(provider);
+    client.generate_rsa_sign_key(key_name.clone()).unwrap();
+
+    let signature = client
+        .sign_with_rsa_sha256(key_name.clone(), HASH.to_vec())
+        .unwrap();
+
+    let pub_key = client.export_public_key(key_name).unwrap();
+
+    Ok((client, pub_key, signature))
+}
+
+fn setup_asym_encr(provider: ProviderID, key_name: String) -> Result<(TestClient, Vec<u8>)> {
+    let mut client = TestClient::new();
+    client.set_provider(provider);
+    client
+        .generate_rsa_encryption_keys_rsapkcs1v15crypt(key_name.clone())
+        .unwrap();
+
+    let pub_key = client.export_public_key(key_name).unwrap();
+
+    Ok((client, pub_key))
+}
+
+fn import_and_verify(
+    client: &mut TestClient,
+    provider: ProviderID,
+    key_name: String,
+    pub_key: Vec<u8>,
+    signature: Vec<u8>,
+) {
+    client.set_provider(provider);
+    client
+        .import_rsa_public_key(key_name.clone(), pub_key)
+        .unwrap();
+    client
+        .verify_with_rsa_sha256(key_name, HASH.to_vec(), signature)
+        .unwrap();
+}
+
+fn import_and_encrypt(
+    client: &mut TestClient,
+    provider: ProviderID,
+    key_name: String,
+    pub_key: Vec<u8>,
+) -> Result<Vec<u8>> {
+    client.set_provider(provider);
+    client
+        .import_rsa_public_key_for_encryption(key_name.clone(), pub_key)
+        .unwrap();
+    client.asymmetric_encrypt_message_with_rsapkcs1v15(key_name, PLAINTEXT_MESSAGE.to_vec())
+}
+
+fn verify_encrypt(
+    client: &mut TestClient,
+    provider: ProviderID,
+    key_name: String,
+    ciphertext: Vec<u8>,
+) -> Result<Vec<u8>> {
+    client.set_provider(provider);
+    client.asymmetric_decrypt_message_with_rsapkcs1v15(key_name, ciphertext)
+}
+
+#[test]
+fn tpm_sign_cross() {
+    let key_name = String::from("tpm_sign_cross");
+    let (mut client, pub_key, signature) = setup_sign(ProviderID::Tpm, key_name.clone()).unwrap();
```

Comment:
Small nit and maybe not good for performance but we could put this function in a `for provider in [Tpm, MbedCrypto, Pkcs11]` loop

Reply:
Hmm, yeah, I guess! My only concern was with figuring out what failed and where, but I guess we can put a `format!` in the `expect(...)`